go/vt/vttablet/tabletmanager/vreplication/table_plan_partial.go

```diff
@@ -0,0 +1,209 @@
+		separator = ","
+	}
+	buf.Myprintf(")", tpb.name)
+	return buf.ParsedQuery()
+}
+
+func (tpb *tablePlanBuilder) generatePartialSelectPart(buf *sqlparser.TrackedBuffer, bvf *bindvarFormatter, dataColumns *binlogdatapb.Bitmap) *sqlparser.ParsedQuery {
+	bvf.mode = bvAfter
+	buf.WriteString(" select ")
+	separator := ""
+	for ind, cexpr := range tpb.colExprs {
+		if tpb.isColumnGenerated(cexpr.colName) {
+			continue
+		}
+		if !isBitSet(dataColumns.Cols, ind) {
+			continue
+		}
+		buf.Myprintf("%s", separator)
+		separator = ", "
+		buf.Myprintf("%v", cexpr.expr)
+
+	}
+	buf.WriteString(" from dual where ")
+	tpb.generatePKConstraint(buf, bvf)
+	return buf.ParsedQuery()
+}
+
+func (tpb *tablePlanBuilder) createPartialInsertQuery(dataColumns *binlogdatapb.Bitmap) *sqlparser.ParsedQuery {
+	bvf := &bindvarFormatter{}
+	buf := sqlparser.NewTrackedBuffer(bvf.formatter)
+
+	tpb.generatePartialInsertPart(buf, dataColumns)
+	if tpb.lastpk == nil {
+		// If there's no lastpk, generate straight values.
+		buf.Myprintf(" values ", tpb.name)
+		tpb.generatePartialValuesPart(buf, bvf, dataColumns)
+	} else {
+		// If there is a lastpk, generate values as a select from dual
+		// where the pks < lastpk
+		tpb.generatePartialSelectPart(buf, bvf, dataColumns)
+	}
+	return buf.ParsedQuery()
+}
+
+func (tpb *tablePlanBuilder) createPartialUpdateQuery(dataColumns *binlogdatapb.Bitmap) *sqlparser.ParsedQuery {
+	bvf := &bindvarFormatter{}
+	buf := sqlparser.NewTrackedBuffer(bvf.formatter)
+	buf.Myprintf("update %v set ", tpb.name)
+	separator := ""
+	for i, cexpr := range tpb.colExprs {
+		if cexpr.isPK {
+			continue
+		}
+		if tpb.isColumnGenerated(cexpr.colName) {
+			continue
+		}
+		if int64(i) >= dataColumns.Count {
+			log.Errorf("Ran out of columns trying to generate query for %s", tpb.name.CompliantName())
+			return nil
+		}
+		if !isBitSet(dataColumns.Cols, i) {
+			continue
+		}
+		buf.Myprintf("%s%v=", separator, cexpr.colName)
+		separator = ", "
+		switch cexpr.operation {
+		case opExpr:
+			bvf.mode = bvAfter
+			switch cexpr.colType {
+			case querypb.Type_JSON:
+				buf.Myprintf("%v", cexpr.expr)
+			case querypb.Type_DATETIME:
+				sourceTZ := tpb.source.SourceTimeZone
+				targetTZ := tpb.source.TargetTimeZone
+				if sourceTZ != "" && targetTZ != "" {
+					buf.Myprintf("convert_tz(%v, '%s', '%s')", cexpr.expr, sourceTZ, targetTZ)
+				} else {
+					buf.Myprintf("%v", cexpr.expr)
+				}
+			default:
+				buf.Myprintf("%v", cexpr.expr)
+			}
+		}
+	}
+	tpb.generateWhere(buf, bvf)
+	return buf.ParsedQuery()
+}
+func (tp *TablePlan) getPartialInsertQuery(dataColumns *binlogdatapb.Bitmap) (*sqlparser.ParsedQuery, error) {
+	key := fmt.Sprintf("%x", dataColumns.Cols)
+	ins, ok := tp.PartialInserts[key]
+	if ok {
+		return ins, nil
+	}
+	ins = tp.TablePlanBuilder.createPartialInsertQuery(dataColumns)
+	if ins == nil {
+		return ins, vterrors.New(vtrpcpb.Code_INTERNAL, fmt.Sprintf("Unable to create partial insert query for %s", tp.TargetName))
```

Comment:
Nit, but error messages aren't supposed to use sentence capitalization (because of potential wrapping).